src/passes/ConstantPropagation.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <pass.h>
+#include <wasm.h>
+
+namespace wasm {
+
+typedef std::map<Name, Const *> State;
+
+// Simple data-flow pass that propagates constants through locals.
+// TODO: Create a generic data-flow optimization pass, and make this class
+// derive from it.
+struct ConstantPropagation : public WalkerPass<PreWalker<ConstantPropagation>> {
+  State state;
+
+  // State at the end of block expressions.
+  std::map<Expression *, State> tails;
+
+  // State at the beginning of loop expressions. Although we collect and merge
+  // states at loop headers, we don't actually use them yet.
+  std::map<Expression *, State> heads;
+
+  // Scope stack of all scope expressions.
+  std::vector<Expression *> scope;
+
+  ConstantPropagation() {}
+
+  // Find out which locals are assigned and erase them from the parent state.
+  struct EraseLocals : public WalkerPass<PreWalker<EraseLocals>> {
+    ConstantPropagation &parent;
+    EraseLocals(ConstantPropagation &parent) : parent(parent) {}
+    void visitSetLocal(SetLocal *set) {
+      tryVisit(&set->value);
```

Comment:
This is using the `PreWalker` instead of the usual `WasmWalker`. The `PreWalker` gives you finer grained control of where you can insert logic. For example, in `visitSwitch` I need to do something special after each case node, which is something I can't do using the `WasmWalker`.